File: src/audio/dsp/drive.ts
/**
 * Drive — soft-clip overdrive blending into hard distortion, with a tilt tone
 * control, output level and selectable clipping character (voice). Pure,
 * deterministic, allocation-free hot path.
 *
 * This is the reference core: every effect follows this shape — a params
 * interface, a `setParams` that clamps, a `processInto(l, r, out)` with no
 * allocation, a `process` test convenience, and `reset`.
 *
 * WHY 2x oversampling around the shaper only: hard-edged curves (Hard, Fold)
 * fold their upper harmonics back below Nyquist. Upsampling by linear interp,
 * shaping both subsamples and decimating with a [1,2,1]/4 halfband kills the
 * worst of that aliasing while the (linear) tone/level stages stay at 1x.
 */
import { clamp, DcBlocker, OnePoleLP, Smoother } from './util.ts'

// Slope-matched normalization reference: dividing by tanh(g/2) alone gave ~2.16x
// small-signal gain at drive=0; scaling by tanh(1/2) restores unity there.
const NORM_REF = Math.tanh(0.5)

// Static bias offsets for the asymmetric voices, subtracted so silence maps to
// exactly 0 (signal-dependent DC is handled by the per-channel DcBlocker).
const TUBE_BIAS = 0.22
const TUBE_OFFSET = Math.tanh(TUBE_BIAS)
const GERM_BIAS = 0.25
const GERM_OFFSET = Math.tanh(GERM_BIAS)

const CHAR_MAX = 6 // [Soft, Hard, Tube, Tape, Germ, Si, Fold]
const CHAR_TAPE = 3

export interface DriveParams {
  drive: number // 0..1
  tone: number // 0..1  (dark -> bright tilt)
  level: number // 0..1  output trim
  character?: number // 0..6 voice index; omitted => 0 (Soft, legacy curve)
}

export class Drive {
  private readonly sampleRate: number
  private readonly scratch = new Float64Array(2)
  private readonly driveS: Smoother
  private readonly toneS: Smoother
  private readonly levelS: Smoother
  // raw targets, set per block; smoothers converge per sample in processInto
  private tDrive = 0.4
  private tTone = 0.55
  private tLevel = 0.85
  // one-pole low-pass state per channel for the tone tilt
  private lpL = 0
  private lpR = 0
  // Character crossfade: ~10ms equal-gain fade between the outgoing and
  // incoming voice so automation never clicks. fade==1 means settled on charTo.
  private charFrom = 0
  private charTo = 0
  private charFade = 1
  private readonly fadeStep: number
  // 2x oversampling state: previous input (for the interpolated midpoint
  // subsample) and previous shaped subsample (halfband decimator history).
  private prevInL = 0
  private prevInR = 0
  private osL = 0
  private osR = 0
  // Tape voice HF rolloff; always processed so a crossfade lands on warm state.
  private readonly tapeLpL = new OnePoleLP()
  private readonly tapeLpR = new OnePoleLP()
  // Asymmetric voices (Tube, Germ) rectify — block the resulting DC.
  private readonly dcL = new DcBlocker()
  private readonly dcR = new DcBlocker()

  constructor(sampleRate: number) {
    this.sampleRate = Number.isFinite(sampleRate) && sampleRate > 0 ? sampleRate : 44100
    this.driveS = new Smoother(this.sampleRate, 0.02, 0.4)
    this.toneS = new Smoother(this.sampleRate, 0.02, 0.55)
    this.levelS = new Smoother(this.sampleRate, 0.02, 0.85)
    this.fadeStep = 1 / (0.01 * this.sampleRate)
    this.tapeLpL.setCutoff(this.sampleRate, 6000)
    this.tapeLpR.setCutoff(this.sampleRate, 6000)
  }

  setParams({ drive, tone, level, character = 0 }: DriveParams): void {
    this.tDrive = clamp(drive, 0, 1)
    this.tTone = clamp(tone, 0, 1)
    this.tLevel = clamp(level, 0, 1)
    const c = Math.round(clamp(character, 0, CHAR_MAX))
    if (c !== this.charTo) {
      if (this.charFade < 1 && c === this.charFrom) {
        // reverse an in-flight fade in place — no jump at all
        this.charFrom = this.charTo
        this.charFade = 1 - this.charFade
      } else {
        // retarget: the (possibly partial) old destination becomes the source
        this.charFrom = this.charTo
        this.charFade = 0
      }
      this.charTo = c
    }
  }

  /** Memoryless base curves; u is the pre-gained input, output roughly [-1,1]. */
  private shapeChar(c: number, u: number): number {
    switch (c) {
      case 1: // Hard — brick clip edge
        return u < -1 ? -1 : u > 1 ? 1 : u
      case 2: // Tube — gentle asymmetric bias => even harmonics
        return Math.tanh(u + TUBE_BIAS) - TUBE_OFFSET
      case 3: // Tape — compressive knee (HF rolloff is blended in post)
        return u / (1 + Math.abs(u))
      case 4: // Germ — saggy asymmetric fuzz knee (generic germanium flavour)
        return Math.tanh(u * (u >= 0 ? 0.8 : 1.3) + GERM_BIAS) - GERM_OFFSET
      case 5: {
        // Si — tighter symmetric fuzz clip
        const v = u * 1.5
        return v / Math.sqrt(1 + v * v)
      }
      case 6: {
        // Fold — reflect peaks back down (triangle fold, identity on [-1,1])
        let t = (u + 1) * 0.25
        t -= Math.floor(t)
        return 1 - Math.abs(4 * t - 2)
      }
      default: // Soft — the legacy tanh curve, bit-compatible at character=0
        return Math.tanh(u)
    }
  }

  /** Crossfaded shaper (both voices evaluated only while a fade is running). */
  private shapeXf(u: number): number {
    const f = this.charFade
    const b = this.shapeChar(this.charTo, u)
    if (f >= 1) return b
    return this.shapeChar(this.charFrom, u) * (1 - f) + b * f
  }

  processInto(left: number, right: number, out: Float64Array): void {
    const drive = this.driveS.process(this.tDrive)
    const tone = this.toneS.process(this.tTone)
    const level = this.levelS.process(this.tLevel)

    // advance the ~10ms character crossfade
    let fade = this.charFade
    if (fade < 1) {
      fade += this.fadeStep
      if (fade >= 1) {
        fade = 1
        this.charFrom = this.charTo
      }
      this.charFade = fade
    }

    // gain scales pre-distortion drive 1..~40; normalize so the small-signal
    // gain is 1 at drive=0 (g=1): the extra NORM_REF factor cancels the
    // built-in makeup gain the old form introduced. The same post-scale also
    // level-compensates the hotter voices (Fold especially).
    const g = 1 + drive * 39
    const k = NORM_REF / Math.tanh(g * 0.5 + 1e-6)

    // tone: interpolate one-pole LP cutoff coefficient by brightness
    const cutoff = 400 + tone * 12000
    const coeff = Math.exp((-2 * Math.PI * cutoff) / this.sampleRate)

    // Guard non-finite input: a single NaN sample would latch the lp state
    // forever (feedback never clears), silencing the whole rack until reset().
    const inL = Number.isFinite(left) ? left : 0
    const inR = Number.isFinite(right) ? right : 0

    // 2x oversampled shaping: subsample A is the linear-interp midpoint of the
    // previous and current input, subsample B is the current input. Decimate
    // with a causal [1,2,1]/4 halfband (previous B, A, B).
    const sAL = k * this.shapeXf(0.5 * (this.prevInL + inL) * g)
    const sBL = k * this.shapeXf(inL * g)
    const sAR = k * this.shapeXf(0.5 * (this.prevInR + inR) * g)
    const sBR = k * this.shapeXf(inR * g)
    let l = 0.25 * this.osL + 0.5 * sAL + 0.25 * sBL
    let r = 0.25 * this.osR + 0.5 * sAR + 0.25 * sBR
    this.osL = sBL
    this.osR = sBR
    this.prevInL = inL
    this.prevInR = inR

    // Tape voice: mild HF rolloff that deepens with drive. Blending filtered
    // against raw keeps every other voice bit-transparent while the filters
    // stay warm for a click-free crossfade into Tape.
    const tapeLpOutL = this.tapeLpL.process(l)
    const tapeLpOutR = this.tapeLpR.process(r)
    const wTape =
      (this.charFrom === CHAR_TAPE ? 1 - fade : 0) + (this.charTo === CHAR_TAPE ? fade : 0)
    if (wTape > 0) {
      const amt = wTape * (0.25 + 0.75 * drive)
      l += (tapeLpOutL - l) * amt
      r += (tapeLpOutR - r) * amt
    }

    // DC-compensate the asymmetric voices (transparent ~5 Hz corner otherwise).
    l = this.dcL.process(l)
    r = this.dcR.process(r)

    this.lpL = l * (1 - coeff) + this.lpL * coeff
    this.lpR = r * (1 - coeff) + this.lpR * coeff
    // Flush denormals (no FTZ in JS) to avoid CPU spikes as the lp state decays.
    if (this.lpL < 1e-20 && this.lpL > -1e-20) this.lpL = 0
    if (this.lpR < 1e-20 && this.lpR > -1e-20) this.lpR = 0
    // blend darkened (lp) with bright (raw) by tone
    l = this.lpL * (1 - tone) + l * tone
    r = this.lpR * (1 - tone) + r * tone

    out[0] = l * level
    out[1] = r * level
  }

  process(left: number, right: number): [number, number] {
    this.processInto(left, right, this.scratch)
    return [this.scratch[0], this.scratch[1]]
  }

  reset(): void {
    this.lpL = 0
    this.lpR = 0
    this.prevInL = 0
    this.prevInR = 0
    this.osL = 0
    this.osR = 0
    this.charFrom = this.charTo
    this.charFade = 1
    this.tapeLpL.reset(0)
    this.tapeLpR.reset(0)
    this.dcL.reset()
    this.dcR.reset()
  }
}

File: native-companion/src/dsp/drive.rs
//! Overdrive with soft / hard / tube / tape / germ / silicon / fold voices.
//!
//! Ported from the browser worklet's drive core (see the design doc). MVP runs
//! at 1× (no oversampling — acceptable for a drive at 48 kHz); the tone control
//! and per-voice normalization match the browser. Asymmetric voices are
//! DC-blocked. Allocation-free; every output is finite-guarded.

use super::{flush, DcBlocker, OnePole};

/// Voice count (character index 0..=6).
const VOICES: u8 = 6;

pub struct Drive {
    drive: f32,
    tone: f32,
    level: f32,
    character: u8,
    tone_lp: [OnePole; 2],
    dc: [DcBlocker; 2],
}

impl Drive {
    pub fn new(sr: f32) -> Self {
        Drive {
            drive: 0.4,
            tone: 0.55,
            level: 0.85,
            character: 0,
            tone_lp: [OnePole::new(), OnePole::new()],
            dc: [DcBlocker::new(sr), DcBlocker::new(sr)],
        }
    }

    pub fn set_params(&mut self, drive: f32, tone: f32, level: f32, character: u8, sr: f32) {
        self.drive = drive.clamp(0.0, 1.0);
        self.tone = tone.clamp(0.0, 1.0);
        self.level = level.clamp(0.0, 1.0);
        self.character = character.min(VOICES);
        let cutoff = 400.0 + self.tone * 12000.0;
        for lp in &mut self.tone_lp {
            lp.set_cutoff(cutoff, sr);
        }
    }

    pub fn reset(&mut self) {
        for lp in &mut self.tone_lp {
            lp.reset();
        }
        for dc in &mut self.dc {
            dc.reset();
        }
    }

    /// Memoryless waveshaper for the current voice.
    fn shape(&self, u: f32) -> f32 {
        match self.character {
            0 => u.tanh(),                            // Soft
            1 => u.clamp(-1.0, 1.0),                  // Hard
            2 => (u + 0.22).tanh() - 0.22_f32.tanh(), // Tube
            3 => u / (1.0 + u.abs()),                 // Tape
            4 => {
                // Germanium: asymmetric bias.
                let a = if u >= 0.0 { 0.8 } else { 1.3 };
                (u * a + 0.25).tanh() - 0.25_f32.tanh()
            }
            5 => {
                // Silicon
                let v = 1.5 * u;
                v / (1.0 + v * v).sqrt()
            }
            _ => fold(u), // Fold
        }
    }

    #[inline]
    pub fn process(&mut self, l: f32, r: f32) -> (f32, f32) {
        (self.channel(0, l), self.channel(1, r))
    }

    #[inline]
    fn channel(&mut self, i: usize, x: f32) -> f32 {
        let g = 1.0 + self.drive * 39.0;
        // Small-signal normalization so louder drive doesn't just get louder.
        let k = 0.5_f32.tanh() / (0.5 * g).tanh();
        let mut y = self.shape(x * g) * k;
        if self.character == 2 || self.character == 4 {
            y = self.dc[i].process(y);
        }
        let lp = self.tone_lp[i].lp(y);
        let toned = lp * (1.0 - self.tone) + y * self.tone;
        flush(toned * self.level)
    }
}

/// Triangle wave-folder, bounded by an iteration cap plus a final clamp so a
/// hostile input can never loop unbounded or leave the [-1, 1] range.
fn fold(x: f32) -> f32 {
    let mut v = x;
    for _ in 0..8 {
        if v > 1.0 {
            v = 2.0 - v;
        } else if v < -1.0 {
            v = -2.0 - v;
        } else {
            break;
        }
    }
    v.clamp(-1.0, 1.0)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn drive_with(character: u8) -> Drive {
        let mut d = Drive::new(48000.0);
        d.set_params(0.6, 0.5, 0.8, character, 48000.0);
        d
    }

    #[test]
    fn silence_stays_silence_for_all_voices() {
        for v in 0..=VOICES {
            let mut d = drive_with(v);
            for _ in 0..256 {
                let (l, r) = d.process(0.0, 0.0);
                assert!(l.abs() < 1e-6 && r.abs() < 1e-6, "voice {v} leaked");
            }
        }
    }

    #[test]
    fn full_scale_stays_finite_and_bounded_for_all_voices() {
        for v in 0..=VOICES {
            let mut d = drive_with(v);
            for i in 0..1000 {
                let x = if i % 2 == 0 { 1.0 } else { -1.0 };
                let (l, r) = d.process(x, x);
                assert!(l.is_finite() && r.is_finite(), "voice {v} non-finite");
                // level<=1 and shapers are bounded ~1, so output stays sane.
                assert!(l.abs() <= 2.0, "voice {v} runaway {l}");
            }
        }
    }

    #[test]
    fn param_extremes_are_finite() {
        for &(dr, tn, lv) in &[(0.0, 0.0, 0.0), (1.0, 1.0, 1.0), (1.0, 0.0, 1.0)] {
            for v in 0..=VOICES {
                let mut d = Drive::new(44100.0);
                d.set_params(dr, tn, lv, v, 44100.0);
                let (l, _) = d.process(0.9, -0.9);
                assert!(l.is_finite());
            }
        }
    }

    #[test]
    fn deterministic() {
        let run = || {
            let mut d = drive_with(2);
            let mut acc = 0.0f32;
            for i in 0..500 {
                let x = ((i as f32) * 0.03).sin();
                acc += d.process(x, x).0;
            }
            acc
        };
        assert_eq!(run(), run());
    }

    #[test]
    fn drive_adds_harmonics_raises_rms() {
        // A hard-driven sine should have more energy than a clean one at the
        // same input (soft clip lifts low-level, adds harmonics).
        let mut clean = Drive::new(48000.0);
        clean.set_params(0.0, 1.0, 1.0, 0, 48000.0);
        let mut dirty = Drive::new(48000.0);
        dirty.set_params(1.0, 1.0, 1.0, 0, 48000.0);
        let (mut ec, mut ed) = (0.0f32, 0.0f32);
        for i in 0..2000 {
            let x = ((i as f32) * 0.05).sin() * 0.2;
            ec += clean.process(x, x).0.powi(2);
            ed += dirty.process(x, x).0.powi(2);
        }
        assert!(ed > ec, "driven energy {ed} not above clean {ec}");
    }
}
